Content.Server/Stains/StainsSystem.cs

```diff
@@ -0,0 +1,111 @@
+using Content.Shared.Chemistry.EntitySystems;
+using Content.Server.DoAfter;
+using Content.Server.Fluids.EntitySystems;
+using Content.Shared.DoAfter;
+using Content.Shared.Stains;
+using Content.Shared.Verbs;
+using Robust.Shared.Prototypes;
+using Content.Shared.Inventory;
+using Content.Shared.FixedPoint;
+
+namespace Content.Server.Stains;
+
+public sealed class StainsSystem : SharedStainsSystem
+{
+    [Dependency] private readonly SolutionContainerSystem _solutionContainer = default!;
+    [Dependency] private readonly IPrototypeManager _prototype = default!;
+    [Dependency] private readonly DoAfterSystem _doAfter = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly PuddleSystem _puddle = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<StainableComponent, InventoryRelayedEvent<SolutionSpilledEvent>>(OnSolutionSpilled);
+        SubscribeLocalEvent<StainableComponent, SolutionSpilledEvent>(OnSolutionSpilled);
+
+        SubscribeLocalEvent<StainableComponent, SolutionChangedEvent>(OnSolutionChanged);
+        SubscribeLocalEvent<StainableComponent, GetVerbsEvent<AlternativeVerb>>(AddSqueezeVerb);
+        SubscribeLocalEvent<StainableComponent, SqueezeDoAfterEvent>(OnSqueezeDoAfter);
+    }
+
+    private void OnSolutionSpilled(EntityUid uid, StainableComponent component, InventoryRelayedEvent<SolutionSpilledEvent> @event)
+    {
+        OnSolutionSpilled(uid, component, ref @event.Args);
```

Comment:
Ensuring that this evenly distributes the reagents will be harder. It might be possible to add a field to the InventoryRelayedEvent indicating how many things are getting relayed to (and which iteration the current event is). That could be used to distribute the reagents pretty well.